swift/apple/FirezoneKit/Sources/FirezoneKit/Views/macOSAlert.swift

```diff
@@ -160,37 +148,67 @@ struct macOSAlert { // swiftlint:disable:this type_name
         Extension Identifier: dev.firezone.firezone.network-extension
         """
 
-        // Code 11
-      case .requestCanceled:
-        // This will happen if the user cancels
-        return nil
+      // Code 11
+    case .requestCanceled:
+      // This will happen if the user cancels
+      return """
+        You must enable the FirezoneNetworkExtension System Extension in System Settings to continue. Until you do,
+        all functionality will be disabled.
+      """
 
-        // Code 12
-      case .requestSuperseded:
-        // This will happen if the user repeatedly clicks "Enable ..."
-        return """
+      // Code 12
+    case .requestSuperseded:
+      // This will happen if the user repeatedly clicks "Enable ..."
+      return """
         You must enable the FirezoneNetworkExtension System Extension in System Settings to continue. Until you do,
         all functionality will be disabled.
 
         For more information and troubleshooting, please contact your administrator.
         """
 
-        // Code 13
-      case .authorizationRequired:
-        // This happens the first time we try to install the system extension.
-        // The user is prompted but we still get this.
-        return nil
+      // Code 13
+    case .authorizationRequired:
+      // This happens the first time we try to install the system extension.
+      // The user is prompted but we still get this.
+      return nil
 
-      @unknown default:
-        return "\(error)"
-      }
-    }()
+    @unknown default:
+      return "\(self)"
+    }
   }
+}
 
-  // Error (fallback case)
-  private static func userMessage(for error: Error) -> String? {
-    return "\(error)"
+// Sometimes errors are passed to us as NSError objects, the Objective-C variant
+extension NSError: UserFriendlyError {
+  func userMessage() -> String? {
+    switch domain {
+    case NEVPNErrorDomain:
+      let code = NEVPNError.Code(rawValue: self.code)! // SAFETY: These are the same error type
+      let err = NEVPNError(code)
+      return err.userMessage()
+    case OSSystemExtensionErrorDomain:
+      let code = OSSystemExtensionError.Code(rawValue: self.code)! // SAFETY: These are the same error type
+      let err = OSSystemExtensionError(code)
+      return err.userMessage()
+    default:
+      return "\(self)"
+    }
+  }
+}
+
+@MainActor
+struct macOSAlert { // swiftlint:disable:this type_name
+  static func show(for error: Error) {
+    let message = (error as UserFriendlyError).userMessage() ?? "\(error)"
```

Comment:
So this means try to cast and if cast fails, fallback to printing the raw error?

Would it be better to define the parameter of the function to enforce at compile-time that it needs to implement the `UserFriendlyError` protocol? That seems more robust because you'll get a compile error any time you want to show an alert for something that doesn't implement the protocol.

Reply:
Well we need this to handle any error. Since we bubble up errors from external frameworks through here, we won't know in advance what they can be.

We picked a couple error types to make friendlier here because they are likely to happen if the user doesn't enable permissions.

Reply:
And? Just implement the protocol for these errors. Or make a separate function that shows a generic error.

The type-system is trying to help you here by flagging all the parts of the code where you are not displaying a user-friendly error dialog.

Reply:
We don't know what errors could be raised. Ther's potentially dozens or hundreds from the various calls from the various libraries we use.

Reply:
Or is the issue that when catching the errors, we don't know their type?

Reply:
Ah, I see what you mean - implement the protocol for `Error` and `NSError`?

Reply:
Correct we don't know in advance what error types are caught

Reply:
> We don't know what errors could be raised. Ther's potentially dozens or hundreds from the various calls from the various frameworks we use.

Ah I thought that upon catch, the compiler enumerates what errors it may through but I guess it doesn't?

Reply:
> Ah, I see what you mean - implement the protocol for `Error` and `NSError`?

That is also an option yeah. What I am planning to do in the Rust GUI client is: Show a generic error like "Oh, we've encountered an expected error. If this persists, contact your administrator." but log the detailed error to Sentry.